modules/GenericDataParser/src/server/generic_data_parser.R

```diff
@@ -3953,6 +4063,133 @@ getFormatParameters <- function(rawOnlyFormat, customFormatSettings, inputFormat
   return(o)
 }
 
+getProtocolEndpointData <- function(protocol) {
+  # Collects all the endpoint values in the protocol object and returns a data frame of them
+
+  # First, we create an empty dataframe for the endpoint data associated with the protocol 
+  protocolEndpointDataFrame <- data.frame(
+    columnName=character(),
+    units=character(),
+    dataType=character()
+  )
+
+  #go through the protocol data to wrangle the endpoint data into the dataframe
+  for (lsState in protocol$lsStates) {
+    if (lsState[['lsKind']] == "data column order") {
+
+      #if the name/units/data type cant be found, submit a NA
+      columnNameEntry = NA
+      unitsEntry = NA
+      dataTypeEntry = NA
+
+      for (value in lsState[['lsValues']]) {
+        #if the lsValue is ignored, skip it 
+        if (value[['ignored']] == FALSE) {
+
+          #depending on what the lsKind is, record the data
+          if (value[['lsKind']] == "column name") {
+            columnNameEntry = value[['stringValue']]
+          } else if (value[['lsKind']] == "column units") {
+            unitsEntry = value[['stringValue']]
+          } else if (value[['lsKind']] == "column type") {
+            dataTypeEntry = value[['stringValue']]
+          }
+          
+        }
+      }
+
+      #add the endpoint to the dataframe
+      dataEntry <- list(columnName = columnNameEntry, units = unitsEntry, dataType= dataTypeEntry)
+      protocolEndpointDataFrame = rbind(protocolEndpointDataFrame, dataEntry, stringsAsFactors=FALSE)
+
+    }
+  }
+
+  return(protocolEndpointDataFrame)
+}
+
+validateExperimentColumns <- function(selColumnOrderInfo, protocolEndpointDataFrame) {
+  # Checks if the experiment columns names/units/data type are found in the protocol endpoint data
+  # Raises errors if an experiment column is not valid 
+  # Only used for protocols with strict endpoint enabled
+
+  for (experimentRowNum in seq(1, nrow(selColumnOrderInfo))) {
+    experimentRowData <- selColumnOrderInfo[experimentRowNum,]
+    experimentRowName = experimentRowData$valueKind
+    experimentRowUnits = experimentRowData$Units
+    experimentRowDataType = experimentRowData$valueType
+    
```

Comment:
We should check concentration and conc units too

Reply:
Should we have additional logic to only validate if conc/conc units are enabled in the conf?

Reply:
That's a great question. Yes, we should add some logic for that case. ACAS is configured to not show concentration in the UI, and someone attempts to upload data that has concentrations into a strict match protocol, we should at the very least warn them. Something along the lines of:

```
The column you're uploading {column name} at concentration {conc} {conc units} cannot be fully validated against protocol {protocol name} because this ACAS server is not configured to track concentration in the Protocol Endpoints. To enable complete checking including concentration, please contact your administrator and ask them to enable "client.protocol.endpointManager.showConcentration"
```